# UI for the Proteomic analysis Shiny app
library(shinythemes)

dataSets <- c("LUMOS","timsTOF")

navbarPage("CLL proteome explorer",
           theme = shinythemes::shinytheme("flatly"),
           tabPanel("Transcriptomic assocation",
                    p("Welcome to the CLL proteome explorer app. This interactive shiny app will help you explore the CLL proteomic dataset and its associations to other multi-omics datasets \
                      included in the publication 'The Protein Landscape of Chronic Lymphocytic Leukemia (CLL)' by Meier-Abt and Lu et al.", style = "color:darkred; font-weight:500; font-size:15px"),
                    titlePanel("Visualize correlations between protein expressions and RNA expressions"),
                    sidebarLayout(
                      sidebarPanel(radioButtons("seleSet","Dataset", dataSets, selected = "LUMOS"),
                                   bsTooltip("seleSet", "Select a platform from which the dataset was acquired.", placement = "bottom", trigger = "hover",
                                             options = NULL),
                                   checkboxInput("ifNoSex","Remove proteins on sex chromosomes"),
                                   bsTooltip("ifNoSex", "Removed proteins encoded by genes on sex chromosomes for all downstream analyses (including analyses in other panels)", placement = "bottom", trigger = "hover",
                                             options = NULL),
                                   textOutput("numAllSample"),
                                   textOutput("numProt"),
                                   hr(),
                                   radioButtons("ifImpute","Impute missing values",c("no","yes"),"no", inline = TRUE),
                                   bsTooltip("ifImpute", "Whether to imput missing values in the dataset using QRILC method for all downstream analyses (including analyses in other panels)", placement = "bottom", 
                                             trigger = "hover",
                                             options = NULL),
                                   hr(),
                                   h5(strong('Filter proteins by the protein-mRNA correlations')),
                                   h6("(This filter only affects the results shown in this panel)"),
                                   sliderInput("coefRange","Coefficient range", -1,1,value = c(0,1), step = 0.05),
                                   textInput("fdrRange", "FDR cut-off", value = 0.1)),
                      mainPanel(DT::dataTableOutput("rnaTab"),
                                hr(),
                                column(h6("A histogram showing the distribution of protein-mRNA correlations coefficients."),
                                  plotOutput("coefPlot", height = 300, width = 350),width=6),
                                column(h6("A scatter plot showing the correlation between protein and mRNA expressions of the selected gene in the above table."),
                                  plotlyOutput("rnaPlot",height = 400, width = 400),width=6))
                    )),
           
           tabPanel("Genomic association",
                    p("In this panel, you can identify and visualize differentially expressed (DE) proteins related to a genomic feature 
                      (gene mutation or copy number variation) or demographic feature (sex). The DE list can be used for analyses (e.g enrichment analysis) in other panels.", 
                      style = "color:darkred; font-weight:500; font-size:15px"),
                    titlePanel("Identify proteins associated with a specified feature"),
                    sidebarLayout(
                      sidebarPanel(
                        radioButtons("seleTestModel","Select a method for DE test",
                                     c("Limma","proDA")), 
                        bsTooltip("seleTestModel", "The Limma method uses moderated t-test to identify DE proteins. Samples with missing values will be ignored in the Limma method. proDA uses a generalized linear model accounting for the missing values by a probalistic dropout model. Limma is significantly faster than proDA ", placement = "bottom", 
                                  trigger = "hover",
                                  options = NULL),
                        checkboxInput("subsetBox", "Subset samples", value= FALSE),
                        bsTooltip("subsetBox", "Perform DE analysis only in a subset of the samples", placement = "bottom", 
                                  trigger = "hover",
                                  options = NULL),
                        conditionalPanel(condition = "input.subsetBox == true",
                                         uiOutput("sampleSubsetBox"),
                                         uiOutput("statusSubsetBox")),
                        uiOutput("featureBox"),
                        uiOutput("blockingBox"),
                        bsTooltip("blockingBox", "Due to the strong impact of trisomy12 and IGHV status on the global protein expression pattern, blocking for trisomy12 and/or IGHV status using a multivariate model can be used to account for potential confounding effect.", placement = "bottom", 
                                  trigger = "hover",
                                  options = NULL),
                        actionButton("calcCorr",label = "Run DE test"),
                        h5(strong("Samples used for testing")),
                        h6("For genomic features, '1' indicates the group of samples with the presence of the selected gene \
                           mutation or copy number variation, while '0' indicates the group without the gene mutation or copy number variations."),
                        tableOutput("infoBinary"),
                        hr(),
                        h5(strong("Filter results")),
                        h6("This filter also affects the analyses in other panels that use the list of differentially expressed proteins generated in this panel."),
                        textInput("pFilter","P-value cutoff", value = 0.01),
                        checkboxInput("ifAdjusted","Use adjusted P-values (BH method)", value = FALSE),
                        textInput("fcFilter","log2(fold change) cutoff", value = 0)
                        ),
                      mainPanel(
                        tags$style(type="text/css",
                                   ".shiny-output-error { visibility: hidden; }",
                                   ".shiny-output-error:before { visibility: hidden; }"
                        ),
                        DT::dataTableOutput("DEtab"),
                        downloadLink('downloadTable', 'Download current table'),
                        plotlyOutput("plot1", width = paste0(500,"px"), height = paste0(400,"px"))
                      ))),
           

           tabPanel("Enrichment analysis",
                    titlePanel("Gene set enrichment analysis"),
                    p("In this panel, you can perform gene set enrichment analysis on the list of the DE proteins generated in the [Genomic association] panel, to investigate the potential pathway activity changes.", 
                      style = "color:darkred; font-weight:500; font-size:15px"),
                    sidebarLayout(
                      sidebarPanel(
                        radioButtons("enrichMethod","Select enrichment method", c("PAGE","GSEA"), inline = TRUE),
                        bsTooltip("enrichMethod", "Either the Parametric Analysis of Gene Set Enrichment (PAGE) method, developed by Kim et al., 2005, or Gene Set Enrichment Analysis (GSEA), developed at Broad Institute can be used for the enrichment analysis. PAGE is faster and less stringent than GSEA.", placement = "bottom", 
                                  trigger = "hover",
                                  options = NULL),
                        conditionalPanel(condition = "input.enrichMethod == 'GSEA'",
                                         numericInput("permNum","Permutation number",value=100,min = 10, max = 10000)),
                        radioButtons("statType", "Statistic used for ranking:", c("log2FC","stat"), inline = TRUE),
                        bsTooltip("statType", "Using either fold change or t-statistics for ranking the proteins. Enrichment results may be slightly different.", placement = "bottom", trigger = "hover",
                                  options = NULL),
                        selectInput("sigSet","Select a geneset collection",list.files(path = "msigs/",pattern = "\\.gmt$"), selectize = FALSE, 
                                    selected = "h.all.v5.1.symbols.gmt", size = 8),
                        bsTooltip("sigSet", "Select a geneset collection downloaded from the Broad MSigDB database for the enrichment analysis.", placement = "bottom", trigger = "hover",
                                  options = NULL),
                        numericInput("sigLevel", label = "P value cut-off for enrichment results", min = 0, max = 1, value = 0.05),
                        checkboxInput("ifFDR", label = "use FDR (BH method)", value = FALSE),
                        actionButton("Run", "Start analysis"),
                        bsTooltip("Run", "Enrichment analysis will take the proteins listed in the 'Genomic association panel' as the input list.", placement = "bottom", trigger = "hover",
                                  options = NULL),width = 3),
                      mainPanel(span(textOutput("errMsg"), style="color:red"),
                                uiOutput("downloadUI1"),
                                uiOutput("tableLegendUI"),
                                DT::dataTableOutput("enrichTab"),
                                column(DT::dataTableOutput("geneTab"),
                                       uiOutput("downloadUI2"),width = 6),
                                column(plotlyOutput("plot2",height = 500, width = 600),width =6)
                      )
                    )),
           tabPanel("Heatmap",
                    titlePanel("Expression heatmap of differentially expressed proteins"),
                    p("In this panel, you can plot the heatmap of the expression of a list of proteins, specificed in the [Plotting paramters] section to visuaize the protein expression variations among samples", 
                      style = "color:darkred; font-weight:500; font-size:15px"),
                    sidebarLayout(
                      sidebarPanel(
                        radioButtons("chooseType",label = "Proteins to plot", 
                                     choices = c("Top variant", "Differentially expressed"),
                                     selected = "Top variant"), 
                        bsTooltip("chooseType", "Select the [Top variant] option to plot top N proteins with the highest variation of expression or select the [Differentially expressed] option to plot a list of DE proteins generated in the [Genomic association] panel", 
                                  placement = "bottom", trigger = "hover", options = NULL),
                        conditionalPanel(condition = "input.chooseType =='Top variant'",
                                         numericInput("numGenes","Number of proteins (top N)", value = 100)),
                       
                        br(),
                        h5(strong("Plotting parameters")),
                        checkboxInput("ifHCcol", "Perform hierarchical clustering on columns", value = FALSE),
                        checkboxInput("ifSet", "Only proteins in selected geneset", value = FALSE),
                        bsTooltip("ifSet", "Plot the heatmap only for genes in a selected gene set", 
                                  placement = "bottom", trigger = "hover", options = NULL),
                        conditionalPanel(condition = "input.ifSet == true",
                                         selectInput("sigSet1","Select geneset database",list.files(path = "msigs/",pattern = "\\.gmt$"), selectize = FALSE, 
                                                     selected = "h.all.v5.1.symbols.gmt", size = 6),
                                         uiOutput("setListBox")),
                        checkboxInput("ifSymbol", "Show gene symbols as row labels", value = TRUE),
                        checkboxInput("ifAddition", "Additional features for column annotation", value = FALSE),
                        bsTooltip("ifAddition", "Select features for column annotation in the heatmap. Selecting multiple features is possible", 
                                  placement = "bottom", trigger = "hover", options = NULL),
                        conditionalPanel(condition = "input.ifAddition == true",
                          uiOutput("colAnnoBox")),
                        h4("Download heatmap as pdf"),
                        numericInput("figWidth", label = "PDF figure width", value = 16),
                        numericInput("figHeight", label = "PDF figure height", value = 15),
                        downloadButton("downHeatmap", "Download")
                      ),
                      mainPanel(span(textOutput("errMsg1"), style="color:red"),
                                plotOutput("plot3",width = 800, height = 800))
                    )),
           tabPanel("Drug response",
                    titlePanel("Identify proteins associated with drug responses"),
                    p("In this panel, you can explore the associations between protein expressions and drug response phenotypes", 
                      style = "color:darkred; font-weight:500; font-size:15px"),
                    sidebarLayout(
                      sidebarPanel(
                        checkboxInput("subsetBox2", "Subset samples", value= FALSE),
                        bsTooltip("subsetBox2", "Perform correlation test only in a subset of the samples", placement = "bottom", 
                                  trigger = "hover",
                                  options = NULL),
                        conditionalPanel(condition = "input.subsetBox2 == true",
                                         uiOutput("sampleSubsetBox2"),
                                         uiOutput("statusSubsetBox2")),
                        uiOutput("blockingBox2"),
                        bsTooltip("blockingBox2", "Due to the strong impact of trisomy12 and IGHV status on the global protein expression pattern, blocking for trisomy12 and/or IGHV status using a multivariate model can be used to account for potential confounding effect.", placement = "bottom", 
                                  trigger = "hover",
                                  options = NULL),
                                  textOutput("numSample"),
                        radioButtons("seleProtOrDrug","", c("Select a drug","Select a protein"), selected = "Select a drug"),
                        bsTooltip("seleProtOrDrug", "You can identify all proteins that associate with a certain drug responses by choosing [Select a drug] or identify all drug responses associated a certain protein by choosing [Select a protein]", placement = "bottom", 
                                  trigger = "hover",
                                  options = NULL),
                        conditionalPanel(condition = "input.seleProtOrDrug == 'Select a drug'",
                                         uiOutput("seleDrugUI")),
                        conditionalPanel(condition = "input.seleProtOrDrug == 'Select a protein'",
                                         uiOutput("seleProtUI"),
                                         checkboxInput("ifDrugAUC","Use AUC of viability")),
                        uiOutput("concRangeUI"),
                        bsTooltip("concRangeUI", "1 is the highest concentration and 5 is the lowest concentration. If multiple concentrations are selected, the viability values are averaged.", placement = "bottom", trigger = "hover",
                                  options = NULL),
                        hr(),
                        uiOutput("seleFeatureUI"),
                        bsTooltip("seleFeatureUI", "You can color the samples (dots) in the correlation plot by a certain feature", 
                                  placement = "bottom", trigger = "hover",
                                  options = NULL),
                        hr(),
                        h4("Filter results"),
                        h6("(restrict the results shown in the table on the right by using filters below)"),
                        textInput("pFilterDrug","p value cutoff", value = 0.01),
                        checkboxInput("ifAdjustedDrug","Use adjusted p values (BH method)", value = FALSE),
                        textInput("fcFilterDrug","log2FC/coefficient cutoff", value = 0)
                      ),
                      mainPanel(
                        tags$style(type="text/css",
                                   ".shiny-output-error { visibility: hidden; }",
                                   ".shiny-output-error:before { visibility: hidden; }"
                        ),
                        DT::dataTableOutput("DrugTab"),
                        downloadLink('downloadDrugTable', 'Download current table'),
                        plotlyOutput("plotDrug", width = paste0(550,"px"), height = paste0(500,"px"))
                      ))))
